catroid/src/androidTest/java/org/catrobat/catroid/test/ui/BaseExceptionHandlerTest.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Catroid: An on-device visual programming system for Android devices
+ * Copyright (C) 2010-2017 The Catrobat Team
+ * (<http://developer.catrobat.org/credits>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * An additional term exception under section 7 of the GNU Affero
+ * General Public License, version 3, is available at
+ * http://developer.catrobat.org/license_additional_term
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.catrobat.catroid.test.ui;
+
+import android.content.Context;
+import android.content.SharedPreferences;
+import android.preference.PreferenceManager;
+import android.support.test.InstrumentationRegistry;
+import android.test.AndroidTestCase;
+
+import org.catrobat.catroid.ui.BaseExceptionHandler;
+import org.catrobat.catroid.ui.SettingsActivity;
+import org.catrobat.catroid.utils.CrashReporter;
+import org.junit.Before;
+
+import static org.catrobat.catroid.utils.CrashReporter.EXCEPTION_FOR_REPORT;
+
+public class BaseExceptionHandlerTest extends AndroidTestCase {
+
+	private SharedPreferences sharedPreferences;
+	private SharedPreferences.Editor editor;
+
+	@Before
+	public void setUp() throws Exception {
+		super.setUp();
+		Context context = InstrumentationRegistry.getTargetContext();
+		sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
+		editor = sharedPreferences.edit();
+		editor.clear();
+		editor.putBoolean(SettingsActivity.SETTINGS_CRASH_REPORTS, true);
+		editor.commit();
+		CrashReporter.setIsCrashReportEnabled(true);
+		CrashReporter.initialize(context);
+	}
+
+	@Override
+	protected void tearDown() throws Exception {
+		super.tearDown();
+		CrashReporter.setIsCrashReportEnabled(false);
+		editor.clear();
+		editor.commit();
+	}
```

Comment:
`super.tearDown();` should be at the end of the method. It's the same principle as with the [pmd rule Android.CallSuperLast](http://pmd.sourceforge.net/pmd-4.3.0/rules/android.html#CallSuperLast).

Best explained with an example.
```
Parent {
  setUp() {
    database = new Database();
    database.open();
  }

  tearDown() {
    database.close();
    database = null;
  }
}

Child extends Parent {
  setUp() {
    super.setUp();
    database.add(item1);
  }

  tearDown() {
    #1 super.tearDown();
    database.remove(item1);
    #2 super.tearDown();
  }
}
```
If you add `super.tearDown` at position \#1 in `Child.tearDown`, you'll see that the database will be closed and set to null. So ` database.remove(item1);` in the Child will not work. Even if you don't set the database object to null in `Parent.tearDown`, the database has been closed already.

Position \#2 will work perfectly. You remove every resource created/initialized in the Child first and then the Parent will release it's resources.

However, you should never work with a real database in unit tests and therefore shouldn't care about removing the item anyway. It's just the first example which came into my mind and I think it perfectly describes the scenario.